File: star2/bar_graph2.rb
# 自作の関数を創って、解くことも出来ます。

# 関数宣言
# 機能：黒い■と白い□を表示する
# 引数：black_number # 黒い■の数
#       white_number # 白い□の数
# 戻値：なし
def black_white_box(black_number, white_number)
  # （特に指定しなければ）
  # 最後に演算した結果が、関数の戻り値となります。
  "■" * black_number + "□" * white_number + "\n"
end

# 引数が一つだけならば良いですが、
# 複数ある場合には順番がわからなくなることもあります。
# 「キーワード引数」を使うと、順番を気にすることなく、
# 関数を呼び出すことが出来ます。
def black_white_box_with_keyword(black_box: black, white_box: white)
  # （特に指定しなければ）
  # 最後に演算した結果が、関数の戻り値となります。
  "■" * black_box + "□" * white_box + "\n"
  # "■" * black ではなく、
  # "■" * black_box と「キーワード名」を書いていることに注意してください。
end

# 表示させたい棒グラフの値配列
values    = [1, 4, 7, 3, 2]
max_value = values.max

# 値配列から要素を取り出し、繰り返します。
values.each do |value|
  black_number = value             # 黒い■の数
  white_number = max_value - value # 白い□の数

  # black_white_box 関数を呼び出し、結果を表示します。
  print black_white_box(black_number, white_number)
end


# キーワード引数を用いて、メソッド呼び出しを行う例です。
# white, black の順で引数を指定しても、支障なく動作します。
values.each do |value|
  black_number = value             # 黒い■の数
  white_number = max_value - value # 白い□の数
  print black_white_box_with_keyword(white_box: white_number,
                                     black_box: black_number)
end

File: star2/bar_graph2.c
#include <stdio.h>  // 標準入出力関数

// 自作の関数を創って、解くことも出来ます。

// 関数のプロトタイプ(原型)宣言
// 機能：黒い■と白い□を表示する
// 引数：int black_number // 黒い■の数
//       int white_number // 白い□の数
// 戻値：なし
void black_white_box(int black_number, int white_number);

int main(int argc, char const *argv[]) {
  // 変数名は、i でも良いですが、
  // 行の変数であることが分かるようにすると良いです。
  int row;       // 行
  int black_number; // 黒い■の数
  int white_number; // 白い□の数

  // 表示させたい棒グラフの値配列
  int values[] = {1, 4, 7, 3, 2};

  // 値配列の要素数が5なので、5行繰り返せば良いです。
  // 予め配列要素の要素数を求めておくと、より汎用的になります。
  for (row = 0; row < 5; row++) {
    black_number = values[row];       // 黒い■の数
    // 値配列の最大値が7なので、7から引いています。
    // 予め配列要素の最大値を求めておくと、より汎用的になります。
    white_number = 7 - black_number;  // 白い□の数

    // 黒い■と白い□を表示する関数を呼び出し、表示を任せます。
    black_white_box(black_number, white_number);
  }

  return 0;
}

// プロトタイプ宣言を先頭でしているので、
// 自作関数本体を、main関数の後に書くことが出来ます。
void black_white_box(int black_number, int white_number) {
  // C言語では、while(0)が偽となり、while文が終了します。
  // ですので、このようにも書くことが出来ます。
  while (black_number--) {
    printf("■"); // ■を表示します。
  }
  while (white_number--) {
    printf("□"); // ■を表示します。
  }
  printf("\n"); // 改行します。
}
